howso/input_validation.amlg

```diff
@@ -249,4 +249,390 @@
+							(lambda
+								(let
+									(assoc
+										parameter_comment (first (current_value 1))
+										default_value (last (current_value 1))
+										parameter_name (current_index 1)
+										method_name (current_index 2)
+
+										typing_map (null)
+									)
+
+									(assign (assoc
+										typing_map
+											;regex to just capture the {} information, debugger adds more and there can be other comments on the parameters
+											(parse (substr parameter_comment "\\{.+\\}"))
+									))
+
+									;uses typing_map, method_name, and parameter_name. accums to invalid_type_hints
+									#!CheckTypingMap
+									(seq
+										(if (or (= (null) typing_map) (= .nas typing_map))
+											(accum (assoc invalid_type_hints (list (concat method_name "/" parameter_name))))
+										)
+
+										(rewrite
+											(lambda
+												(if (or
+														(and
+															;if a string value that should be a type, ensure it is a valid type
+															(~ "" (current_value))
+															(or
+																(contains_value ["type" "values" "additional_indices"] (current_index))
+																(= "indices" (current_index 1))
+															)
+															(not (contains_value ["boolean" "number" "string" "list" "assoc" "any"] (current_value) ))
+														)
+														(and
+															;if a string value reference, ensure that reference is in the !customTypes
+															(~ "" (current_value))
+															(= "ref" (current_index))
+															(not (contains_index !customTypes (current_value)))
+														)
+														(and
+															;if an assoc (that's not the value of "indices"), it must have a type or ref
+															(~ (assoc) (current_value))
+															(!= "indices" (current_index))
+															(and
+																(not (contains_index (current_value) "type"))
+																(not (contains_index (current_value) "ref"))
+															)
+														)
+													)
+													(accum (assoc invalid_type_hints (list (concat method_name "/" parameter_name))))
+												)
+											)
+											typing_map
+										)
+									)
+
+									;return the typing map with default/optional added
+									(append
+										typing_map
+										(assoc
+											default default_value
+											optional (= default_value (null))
+										)
+									)
+								)
+							)
+							;first item is the assoc of parameter names to comments/default value
+							(first (get_entity_comments (null) (current_index) (true)))
+						)
+					)
+					(get_entity_comments (null) (null) (true))
+				)
+		))
+
+		;also need to check the CustomTypes schemas
+		(map
+			(lambda
+				(call !CheckTypingMap (assoc
+					typing_map (current_value 1)
+					method_name "CustomTypes"
+					parameter_name (current_index 1)
+				))
+			)
+			!customTypes
+		)
+
+		(if (= 0 (size invalid_type_hints))
```

Comment:
can simplify to: `(if (size invalid_type_hints) invalid_type_hints (assign_to_entities ..) )`